### ISSUE TITLE
Support exhaustiveness checking of single enumerable properties

### ISSUE THREAD

**Issue by aiven-anton** [python/mypy#14261]
**Feature**

Support exhaustiveness checking of single boolean and enum properties and attributes of objects.

**Pitch**

The proposal is to improve exhaustiveness checking by supporting properties/attributes on a matched object. While it would be amazing to have infinitely nested support for this, I think we can recognize that there is value in supporting just single-level enumerable attributes.

In this example mypy isn't currently recognizing that all possible variations of `Foo` are handled, and gives an error on the `assert_never()`.

```python
from typing import assert_never

class Sentinel: ...

class Foo:
    prop: bool

val: Foo | Sentinel

match val:
    case Sentinel():
        ...
    case Foo(prop=True):
        ...
    case Foo(prop=False):
        ...
    case no_match:
        assert_never(no_match)  # Currently a false positive here, no_match should be <nothing>.
```

https://mypy-play.net/?mypy=latest&python=3.11&gist=1edcb8f3ea56334662ed2036404563a1

Likewise, if the feature is accepted, we should expect an unreachable error for the third `case` in this example.

```python
from typing import assert_never

class Foo:
    prop: bool

val: Foo

match val:
    case Foo(prop=True):
        ...
    case Foo(prop=False):
        ...
    case Foo():
        ...  # Currently a false negative here, should be unreachable.
```

https://mypy-play.net/?mypy=latest&python=3.11&gist=74b23b6868c0c52056be1ade612b3052

**Comment by erictraut** [comment]
Exhaustiveness checking is implemented using type narrowing. In your first example above, the first case statement (`Sentinel()`) narrows the type of `val` from `Sentinel | Foo` to `Foo`. The next case statement (`Foo(prop=True)`) does not further narrow the the type of `val`. There is no type in the Python type system that represents a `Foo` object whose instance variable `prop` has been proven to be `Literal[False]`.

The way to make this work with exhaustiveness validation is to change your code to the following:

```python
match val:
    case Sentinel():
        ...
    case Foo():
        if val.prop:
            # do something
            ...
        else:
            # do something else
            ...
    case no_match:
        assert_never(no_match)  # Currently a false positive here, no_match should be <nothing>.
```
